File: 350c_bombs.py
#! /usr/bin/python

import sys

def get_x_dir(x):
    if x == 0:
        return None
    return 'R' if x > 0 else 'L'

def get_y_dir(y):
    if y == 0:
        return None
    return 'U' if y > 0 else 'D'

if __name__ == '__main__':
    sys.stdin.readline()
    bombs = []

    d = 0
    for line in sys.stdin.readlines():
        x, y = map(int, line.split())
        if x == 0 or y == 0:
            d += 1
        bombs.append((x, y))

    bombs.sort(lambda a, b: abs(a[0]) + abs(a[1]) - abs(b[0]) - abs(b[1]))

    print len(bombs) * 6 - 2 * d

    for b in bombs:
        x, y = b
        x_dir = get_x_dir(x)
        if x_dir != None:
            print '1 %d %s' % (abs(x), x_dir)
        y_dir = get_y_dir(y)
        if y_dir != None:
            print '1 %d %s' % (abs(y), y_dir)
        print '2'
        # go back to (0, 0)
        if x_dir != None:
            print '1 %d %s' % (abs(x), get_x_dir(-x))
        if y_dir != None:
            print '1 %d %s' % (abs(y), get_y_dir(-y))
        print '3'

File: 350c_bombs.cc
#include <iostream>
#include <utility>
#include <algorithm>
#include <cmath>

using namespace std;

const char UNDEFINED = 'X';
typedef pair<int, int> Bomb;

char get_x_dir(int x) {
    if (x == 0) {
        return UNDEFINED;
    }
    return x > 0 ? 'R' : 'L';
}

char get_y_dir(int y) {
    if (y == 0) {
        return UNDEFINED;
    }
    return y > 0 ? 'U' : 'D';
}

Bomb make_bomb(int x, int y) {
    return make_pair(x, y);
}

bool bomb_cmp (const Bomb a, const Bomb b) {
    return (abs(a.first) + abs(a.second) - abs(b.first) - abs(b.second)) < 0;
}

int main(int argc, char *argv[]) {
    int N;
    cin >> N;
    cin.get();
    Bomb *bomb = new Bomb[N];

    int d = 0;
    for (int i=0; i < N; ++i) {
        int x, y;
        cin >> x;
        cin >> y;
        cin.get();
        if (x == 0 || y == 0) {
            ++d;
        }
        bomb[i] = make_bomb(x, y);
    }

    sort(bomb, bomb + N, bomb_cmp);

    cout << (N * 6 -  2 * d) << endl;
    for (int i=0; i < N; ++i) {
        int x = bomb[i].first;
        int y = bomb[i].second;
        char x_dir = get_x_dir(x);
        if (x_dir != UNDEFINED) {
            printf("1 %d %c\n", abs(x), x_dir);
        }
        char y_dir = get_y_dir(y);
        if (y_dir != UNDEFINED) {
            printf("1 %d %c\n", abs(y), y_dir);
        }
        cout << '2' << endl;
        // go back to (0, 0)
        if (x_dir != UNDEFINED) {
            printf("1 %d %c\n", abs(x), get_x_dir(-x));
        }
        if (y_dir != UNDEFINED) {
            printf("1 %d %c\n", abs(y), get_y_dir(-y));
        }
        cout << '3' << endl;
    }
    delete[] bomb;
    return 0;
}
